Quality of life improvements

alexreardon opened the pull request:
Adding a collection of improvements to `memoize-one`. Still maintaining ie11 support at this stage

### TODO


- [x] Add a `.clear()` function property to memoized functions to allow for cache clearing
- [x] Improving types of the memoized function
- [x] Improving types of `EqualityFn`
- [x] Adding a node script to compare the performance of `memoize-one` against other memoization libraries
- [x] Upgrade all dependencies
- [x] Shift over from TravisCI to Github actions
- [x] Add type tests to ensure types are working as expected
- [x] Add type tests for `EqualityFn`
- [x] update flow types
- [x] Release notes
- [x] README: new performance command
- [x] README: using the `EqualityFn` type
- [x] README: function properties
- [x] README: `.clear()`

### To be decided

- [x] Use `MemoizedFn` type or just inline that type?

**Option 1: Use a `MemoizedFn` type**

(probably makes sense to export `MemoizedFn`)

```ts
export type MemoizedFn<TFunc extends (this: any, ...args: any[]) => any> = {
  clear: () => void;
  (this: ThisParameterType<TFunc>, ...args: Parameters<TFunc>): ReturnType<TFunc>;
};

function memoizeOne<TFunc extends (this: any, ...newArgs: any[]) => any>(
  resultFn: TFunc,
  isEqual: EqualityFn<TFunc> = areInputsEqual,
): MemoizedFn<TFunc> {
/*...*/
}
```

**Option 2: Inline the new type**

```ts
function memoizeOne<TFunc extends (this: any, ...newArgs: any[]) => any>(
  resultFn: TFunc,
  isEqual: EqualityFn<TFunc> = areInputsEqual,
): {
  clear: () => void;
  (this: ThisParameterType<TFunc>, ...args: Parameters<TFunc>): ReturnType<TFunc>;
} {
/*...*/
}
```

Went for *option 1* as it reads nicer, as well as exposing a `MemoizedFn` type that people can more easily use if they want to

- [x] Should this version be a `minor` or a `major`

`major`. On balance I thought this was the least harmful. I know if I did it as a `minor` somebody would flame me in an issue 😅

- [x] Do we need `type:module` in the `package.json`?

No.

Just going to leave it as is for now and can revisit

- [x] Do we add `.length`? https://github.com/alexreardon/memoize-one/pull/18

No.

Code to add the `.length` property:

```ts
Object.defineProperty(memoized, 'length', {
  writable: false,
  enumerable: false,
  configurable: true,
  value: resultFn.length,
});
```

This will `throw` on ie11. I *could* wrap the assignment in a try/catch. Right now this project still supports ie11, and this project is still leaned on super heavily in the JS community. I'm not looking to drop ie11 support just yet.

I am currently thinking not to try to add `.length` (even in a try/catch). I don't want the memoized function to be different in different environments as this might cause some unexpected behaviours 

- [x] Should we update the memoized functions `.name` and/or `.displayName` property? (see #122)

 No.

Adding a `.name` / `.displayName` property to a memoized function does not yield as good of results as I would have liked, and to support ie11 the `.name` property assignment would need to be wrapped in a `try/catch`; potentially further hurting performance.

alexreardon commented on src/memoize-one.ts:
```diff
@@ -1,40 +1,56 @@
-  let lastArgs: unknown[] = [];
-  let lastResult: ReturnType<ResultFn>;
-  let calledOnce: boolean = false;
+export type EqualityFn<TFunc extends (...args: any[]) => any> = (
```
The types of the EqualityFn are now based on the function that it is being used for

alexreardon commented on src/memoize-one.ts:
```diff
@@ -1,40 +1,56 @@
+  lastResult: ReturnType<TFunc>;
+};
+
+type MemoizedFn<TFunc extends (this: any, ...args: any[]) => any> = {
```
Previously we were casting the return value to be `ResultFn`

```ts
return memoized as ResultFn;
```

That is technically not correct as we are not maintaining any properties on the function itself. We now return a function that has the same type as the functions call signature and we are explicitly adding a `.clear` property to that

theKashey commented:
> Do we add .length?

__Yes__. Some libraries, like `redux.connect` use this information.

theKashey commented:
> Adding a .name / .displayName property

Can it be dev-mode only? As it does not make any sense to pass _minified_, _uglified_ and _tersed_ named anywhere.

alexreardon commented:
I'm exploring dev tooling separately (there is a lot more that could be done) so for now i'll stay away from adding `name`

alexreardon commented:
Can we do the `.length` assignment in ie11?

alexreardon commented:
Need to check if this working in ie11:

```ts
// Faking the 'length' property of the function to match the original function
  Object.defineProperty(memoized, 'length', {
    writable: false,
    configurable: true,
    value: resultFn.length,
  });
```

Now to get an ie11 environment! I haven't needed one in a while 😅

alexreardon commented:
It would be good also to confirm if setting .length like this causes any deoptimisations

alexreardon commented:
Adding `.length` seems to have a non-trivial performance hit

<img width="601" alt="Screen Shot 2021-09-28 at 7 35 07 pm" src="https://user-images.githubusercontent.com/2182637/135062691-65a4d7ae-f904-4af0-a60c-b336273b8ba0.png">

alexreardon commented:
I'm currently leaning towards not adding `.length` as it seems to have a non-trivial performance impact

<img width="286" alt="Screen Shot 2021-09-29 at 9 34 30 am" src="https://user-images.githubusercontent.com/2182637/135255028-0e27fa87-7caf-45b0-a220-1420a8c479ef.png">

alexreardon commented:
It turns out adding `.length` doesn't have the original performance drawback I was seeing. I was not setting the property `value` which I think was having a negative impact

<img width="291" alt="Screen Shot 2021-09-29 at 9 14 28 pm" src="https://user-images.githubusercontent.com/2182637/135257921-4326d7fa-7cc3-4c89-a473-bbb561a36e00.png">

Based on these numbers I am happy to start looking at ie11 support

alexreardon commented:
Code to add the `.length` property:

```ts
Object.defineProperty(memoized, 'length', {
	  writable: false,
	  enumerable: false,
	  configurable: true,
	  value: resultFn.length,
});
```

This will `throw` on ie11. I *could* wrap the assignment in a try/catch. Right now this project still supports ie11, and this project is still leaned on super heavily in the JS community. I'm not looking to drop ie11 support just yet.

I am currently thinking not to try to add `.length` (even in a try/catch). I don't want the memoized function to be different in different environments as this might cause some unexpected behaviours

theKashey commented:
Take a look at https://github.com/blakeembrey/arity/blob/master/arity.js, which I am using for a few years and it can be a good fallback for IE11

Why failback - "eval" might not be strict CSP rules compatible and is not proven to work in all environments.

alexreardon commented:
So here is where the rubber hits the road; should this change be a `major` or a `minor`?

### memoized function `type` is now more accurate

*Old: result of the memoized function was being cast to `ResultFn` (the same type as the function that is passed in)*

```ts
declare function memoizeOne<ResultFn extends (this: any, ...newArgs: any[]) => ReturnType<ResultFn>>
  (resultFn: ResultFn, isEqual?: EqualityFn): ResultFn;
```

*New: result of memoized function is more nuanced*

- the returned memoized function is `(this: ThisParameterType<TFunc>, ...args: Parameters<TFunc>): ReturnType<TFunc>` (being clear that this is not the *exact* same thing as `TFunc` (`resultFn`) as we are not preserving any properties on the function (eg `originalFn.foobar`
- return function also has a `.clear()` property *[new API]*

```ts
declare type MemoizedFn<TFunc extends (this: any, ...args: any[]) => any> = {
    clear: () => void;
    (this: ThisParameterType<TFunc>, ...args: Parameters<TFunc>): ReturnType<TFunc>;
};
declare function memoizeOne<TFunc extends (this: any, ...newArgs: any[]) => any>
  (resultFn: TFunc, isEqual?: EqualityFn<TFunc>): MemoizedFn<TFunc>;
```

### `EqualityFn` `type` is now parametrised

*Old: new and last arguments are an array of `any`*

```ts
type EqualityFn = (newArgs: any[], lastArgs: any[]) => boolean;
```

*New: new and last arguments are typed as the parameters to the function memoize memoized 🤩*

```ts
type EqualityFn<TFunc extends (...args: any[]) => any> = (
  newArgs: Parameters<TFunc>,
  lastArgs: Parameters<TFunc>,
) => boolean;
```

alexreardon commented:
> Take a look at https://github.com/blakeembrey/arity/blob/master/arity.js, which I am using for a few years and it can be a good fallback for IE11
> 
> Why failback - "eval" might not be strict CSP rules compatible and is not proven to work in all environments.

This is very creative @theKashey! I think for now it is still best to stay away from this creativity.

- `memoize-one` has never supported maintaining `.length`
- I hope to drop ie11 support soonish, and so this level of creativity would not be needed to add `.length` (could just use `Object.defineProperty`)

theKashey commented:
> memoize-one has never supported maintaining .length

In this case I reckon it will be nice to __define__ this property in dev mode, and emit a warning on access. _Some_ functionality might use `.length` for _some_ purpose, and it will be close to impossible to understand why something is not working after being memoized.

alexreardon commented:
@TrySound what do you think we should do with the existing flowtype? (I want to add a `.clear()` property to the `ResultFn` type)

alexreardon commented:
Thanks @TrySound! I'm trying to figure out how to give the callable property the same signature as ResultFn. I worked it out for typescript, but no success with flow yet. 

I suspect $Call will be needed but I haven't figured it out yet

https://flow.org/en/docs/types/utilities/#toc-call

alexreardon commented:
You can try this out now with `memoize-one@6.0.0-beta.1`

alexreardon commented:
# [Draft] Release notes

## At a glance

🧹 New `.clear()` function to remove the current memoization cache
🏋️‍♀️ Stronger types
🥰 Improved documentation

This release is a `major`, but there are no behaviour or API changes. The `major` is to reflect that some of the `TypeScript` types have been tightened which might cause some consumers `TypeScript` builds to break.

> PSA: `memoize-one` will soon be dropping ie11 support
> [More details](https://github.com/alexreardon/memoize-one/issues/125)

## New: Cache releasing with `.clear()` 🧹

A `.clear()` property is now added to memoized functions to allow you to clear its memoization cache

This is helpful if you want to:

- Clear memory
- Cause the underlying function to be called again without having to change arguments

```ts
import memoizeOne from 'memoize-one';

function add(a: number, b: number): number {
  return a + b;
}

const memoizedAdd = memoizeOne(add);

// first call - not memoized
const first = memoizedAdd(1, 2);

// second call - cache hit (underlying function not called)
const second = memoizedAdd(1, 2);

// 👋 clearing memoization cache
memoizedAdd.clear();

// third call - not memoized (cache was cleared)
const third = memoizedAdd(1, 2);
```

## Improved: memoized function `type`

> There are no API changes to `memoize-one`, this is merely a typing improvement

Our previous `type` for a memoized function was simple:

**Previous**

```ts
export declare type EqualityFn = (newArgs: any[], lastArgs: any[]) => boolean;

declare function memoizeOne<ResultFn extends (this: any, ...newArgs: any[]) => ReturnType<ResultFn>>(resultFn: ResultFn, isEqual?: EqualityFn): ResultFn;
```

A memoized function claimed to be the same type (`ResultFn`) as the original function. This was not true, as `memoize-one` does not copy of any existing object properties on the original function (`TFunc`)

**Updated**

```ts
export declare type MemoizedFn<TFunc extends (this: any, ...args: any[]) => any> = {
  clear: () => void;
  (this: ThisParameterType<TFunc>, ...args: Parameters<TFunc>): ReturnType<TFunc>;
};

declare function memoizeOne<TFunc extends (this: any, ...newArgs: any[]) => any>(
  resultFn: TFunc,
  isEqual?: EqualityFn<TFunc>,
): MemoizedFn<TFunc>;
```

A memoized function returns the same callable signature as the original function (`TFunc` → was `ResultFn`), but it makes it clear that no function object properties on the original function (`TFunc`) are being carried forward. The memoized function also now includes a `.clear()` function object property

> This `type` change has been labelled as a `patch` (fix) as the previous type was not correct. However, you _could_ consider it a `major` given that the new type is a bit narrower than before

## Improved: equality function `type`

> There are no API changes to equality functions, this is merely a typing improvement

**Previous**

```ts
export type EqualityFn = (newArgs: any[], lastArgs: any[]) => boolean;
```

**Current**

```ts
export type EqualityFn<TFunc extends (...args: any[]) => any> = (
  newArgs: Parameters<TFunc>,
  lastArgs: Parameters<TFunc>,
) => boolean;
```

This looks a little scary, but it is pretty neat! It means that you can dramatically improve the type safety of your custom equality functions if you want to.

**If you are not using a custom equality function**

No changes for you!

**If you are using a custom equality function**

Most people will not be impacted!

This type tightening allows you to be a lot stricter with the shape of your functions passed in as equality functions. If you are using generic equality functions such as `lodash.isequal` their types are loose and there is nothing you will need to do. But if you want to write more efficient and typesafe equality functions, you are in for a treat.

**An example of what things looked like in 5.x**

```ts
import memoize, { EqualityFn } from "memoize-one";

type Person = {
  id: string;
  name: string;
};

function invite(person: Person) {
  // This could do something fancy, but just keeping things basic
  console.log("invited:", person.name);
}

// Yuck, we don't know anything about the args
// Note: don't really need the `EqualityFn` type as it is just:
// `type EqualityFn = (newArgs: any[], lastArgs: any[]) => boolean;`
const isEqual: EqualityFn = (newArgs: any[], lastArgs: any[]): boolean => {
  // Yuck #2: we have to cast here
  // We would also be creating a bug if isEqual is used on a function
  // that has no arguments
  const first = newArgs[0] as Person;
  const second = lastArgs[0] as Person;
  return first.id === second.id;
};

const memoized = memoize(invite, isEqual);

const alex: Person = {
  name: "Alex",
  id: "11111"
};

memoized(alex);
// Won't do anything as `alex` has the same id as the last `alex`
memoized(alex);
```

**The same example in 6.x**

```ts
import memoize, { EqualityFn } from "memoize-one";

type Person = {
  id: string;
  name: string;
};

function invite(person: Person) {
  console.log("invited:", person.name);
}

// Yum: we know that newArgs + lastArgs are the tuple `[Person]`
const isEqual: EqualityFn<typeof invite> = ([first], [second]): boolean => {
  return first.id === second.id;
};

const memoized = memoize(invite, isEqual);

const alex: Person = {
  name: "Alex",
  id: "11111"
};

memoized(alex);
// Won't do anything as `alex` has the same id as the last `alex`
memoized(alex);

// When declared inline, our isEqual function has the correct types inferred
const inferred = memoize(invite, function isEqual([first], [second]): boolean {
  return first.id === second.id;
});
```

> There are a few cases where this could cause your `TypeScript` types to start failing, so this change has been listed as a `major`

## Improved: documentation

- Cleanup and audit of examples
- Added documentation for `.clear()`
- Added documentation about function properties and the `.length` property
- Updated performance benchmarks

## Code health

- Now on `TypeScript@4.4.3`
- Moved from TravisCI to Github workflows
- Upgraded `devDependencies`

alexreardon commented:
The current plan is to release 6.x tomorrow 😲
